src/daemon/ca/publishing.rs

```diff
@@ -1,232 +1,1223 @@
+                    }
+                    super::CaEvtDet::KeyRollFinished { resource_class_name } => {
+                        objects.keyroll_finish(resource_class_name)?;
+                    }
+                    super::CaEvtDet::CertificateReceived {
+                        resource_class_name,
+                        rcvd_cert,
+                        ..
+                    } => {
+                        // Update the received certificate if needed. If the URIs changed we may need to re-issue things
+                        objects.update_received_cert(resource_class_name, rcvd_cert)?;
+                        objects.re_issue_if_required(&self.config.issuance_timing, &self.signer)?;
+                    }
+                    super::CaEvtDet::ResourceClassRemoved {
+                        resource_class_name, ..
+                    } => {
+                        objects.remove_class(resource_class_name);
+                    }
+                    super::CaEvtDet::RepoUpdated { contact } => {
+                        objects.update_repo(contact);
+                    }
+                    _ => {}
+                }
+            }
+            Ok(())
+        })
+    }
 }
 
-//------------ ManifestInfo ------------------------------------------------
+impl CaObjectsStore {
+    fn key(ca: &Handle) -> KeyStoreKey {
+        KeyStoreKey::simple(format!("{}.json", ca))
+    }
+
+    fn cas(&self) -> KrillResult<Vec<Handle>> {
+        let cas = self
+            .store
+            .read()
+            .unwrap()
+            .keys(None, ".json")?
+            .iter()
+            .flat_map(|k| {
+                let mut res = None;
+                if let Some(name) = k.name().strip_suffix(".json") {
+                    if let Ok(handle) = Handle::from_str(name) {
+                        res = Some(handle)
+                    }
+                }
+                res
+            })
+            .collect();
+        Ok(cas)
+    }
+
+    /// Get objects for this CA, create a new empty CaObjects if there is none.
+    pub fn ca_objects(&self, ca: &Handle) -> KrillResult<CaObjects> {
+        let key = Self::key(ca);
+
+        match self.store.read().unwrap().get(&key).map_err(Error::KeyValueError)? {
+            None => {
+                let objects = CaObjects::new(ca.clone(), None, HashMap::new());
+                Ok(objects)
+            }
+            Some(objects) => Ok(objects),
+        }
+    }
+
+    pub fn with_ca_objects<F>(&self, ca: &Handle, op: F) -> KrillResult<()>
+    where
+        F: FnOnce(&mut CaObjects) -> KrillResult<()>,
+    {
+        let mut objects = self.ca_objects(ca)?;
+        op(&mut objects)?;
+        self.put_ca_objects(ca, &objects)?;
+        Ok(())
+    }
+
+    pub fn put_ca_objects(&self, ca: &Handle, objects: &CaObjects) -> KrillResult<()> {
+        self.store
+            .write()
+            .unwrap()
```

Comment:
Is `unwrap()` the right thing to do here and not instead `read()?` or am I missing something?